tiatoolbox/annotation/storage.py

```diff
@@ -1007,6 +1051,197 @@ def select_values(
+            >>> store.add(neighbour, "bar")
+            >>> store.nquery((-.5, -.5, .5, .5), distance=2.0)
+            {
+              "foo": {
+                Annotation(POINT (0 0), {'class': 42}): {
+                  "bar": Annotation(POINT (1 1), {'class': 123}),
+                }
+              },
+            }
+
+            Example bounding box query with no neighbours within a
+            distance of 1.0.
+
+            >>> from shapely.geometry import Point
+            >>> from tiatoolbox.annotation.storage import Annotation, SQLiteStore
+            >>> store = SQLiteStore()
+            >>> annotation = Annotation(Point(0, 0), {"class": 42})
+            >>> store.add(annotation, "foo")
+            >>> store.nquery((-.5, -.5, .5, .5), distance=1.0)
+            {"foo": {Annotation(POINT (0 0), {'class': 42}): {}}}
+
+            Example of querying for TILs - lympocytes within 3 units
+            of tumour cells.
+
+            >>> from tiatoolbox.annotation.storage import SQLiteStore
+            >>> store = SQLiteStore("hovernet-pannuke-output.db")
+            >>> tils = store.nquery(
+            ...     where="props['class'] == 1",   # Tumour cells
+            ...     n_where="props['class'] == 0",  # Lymphocytes
+            ...     distance=32.0,  # n_where within 32 units of where
+            ...     mode="point-point",  # Use point to point distance
+            ... )
+
+        """
+        # This is a naive generic implementation which can be overridden
+        # by back ends which can do this more efficiently.
+        if not isinstance(mode, (str, tuple)):
+            raise TypeError("mode must be a string or tuple of strings")
+        if isinstance(mode, str):
+            mode = tuple(mode.split("-"))
+        if mode not in (("box", "box"), ("boxpoint", "boxpoint"), ("poly", "poly")):
+            raise ValueError(
+                "mode must be one of 'box-box', 'boxpoint-boxpoint', or 'poly-poly'"
+            )
+        from_mode, _ = mode
+
+        # Initial selection of annotations to query around
+        selection = self.query(
+            geometry=geometry,
+            where=where,
+        )
+
+        # Query for others within the distance of initial selection
+        result = {}
+        for key, ann in selection.items():
+            geometry = ann.geometry
+            if from_mode == "box":
+                geometry_predicate = "bbox_intersects"
+                min_x, min_y, max_x, max_y = ann.geometry.bounds
+                geometry = Polygon.from_bounds(
+                    min_x - distance,
+                    min_y - distance,
+                    max_x + distance,
+                    max_y + distance,
+                )
+            elif from_mode == "boxpoint":
+                geometry_predicate = "bbox_centre_within"
+                geometry = Polygon.from_bounds(*ann.geometry.bounds).centroid
+                geometry = geometry.buffer(distance)
```

Comment:
the above seems like it could be done more efficiently, as it will end up only using the bounds of the query geometry. The above essentially makes a box, gets its center, makes a bigger circle, then makes another box when the bbox_center_within is calculated. Instead, couldnt we:

```
#outside the for loop, make a box:  
query_box = box(-d/2, -d/2, d/2, d/2)  
#then instead of the above two lines  
bounds = ann.geometry.bounds  
geometry = translate(query_box, (bounds[2]-bounds[0])/2, (bounds[3]-bounds[1])/2)  
```
that would avoid the buffering of the point into a circle, which as it has to represent it as a polygon may be expensive.

Reply:
Just checked and the predicate function (for DictionaryStore) is correct. It is checking if the centroid of the annotation geometry is inside the bounding box of the query geometry as passed to _query.

The mix-up here is that it is using the annotation geometry bounds and the query geometry centoid (which we needlessly make a centroid already here), not the other way around. This is due to how the spatial index functions.

Reply:
The centroid used here is the centroid of the bounds and not the centroid of the geometry for rtree compatibility reasons, hence creating the bounds and then getting the centroid. The ann is unique for each loop iteration here, so cannot be moved outside the loop.

Reply:
So here we get the bounds and then the centroid then buffer instead of simply centroid and then buffer because geometry.centroid != the centre point of the bounding box.

Reply:
However, I do see that maybe this loop case should maybe the same as the above "bbox_intersects" except with the different geometry predict to be equivalent to SQLiteStore.

Reply:
Thanks for raising this! I finally managed to work through it and fix the issue!